Task 3/Task 3.2/Task3.2.1.Library/DynamicArray.cs

```diff
@@ -0,0 +1,281 @@
+        {
+            _array = new T[capacity];
+            Length = 0;
+        }
+
+        /// <summary>
+        /// Конструктор, который в качестве параметра принимает коллекцию, реализующую 
+        /// интерфейс IEnumerable<T>, создаёт массив нужного размера и копирует в него все
+        /// элементы из коллекции.
+        /// </summary>
+        public DynamicArray(IEnumerable<T> source)  // 3
+        {
+            _array = source.ToArray();
+            Length = source.Count();
+        }
+
+
+        /// <summary>
+        /// The method must be called before each addition to the array.
+        /// Increases the capacity to the required value, if needed.
+        /// (The array will grow as well)
+        /// </summary>
+        private void IncreaseCapacity(int requiredValue) // 2*
+        {
+            if (requiredValue <= Capacity)
+                return;
+
+            int newCapacity = Capacity == 0 ? 8 : Capacity;
+
+            while (newCapacity < requiredValue) { newCapacity *= 2; }
+
+            T[] newArray = new T[newCapacity];
+            Array.Copy(_array, newArray, Length);
+            _array = newArray;
+        }
+
+        /// <summary>
+        /// The method shifts the elements of the array, starting from the
+        /// specified position, to the right by the specified number of positions.
+        /// (Increases capacity if needed)
+        /// </summary>
+        private void RightShift(int position, int numberOfPos)
+        {
+            IncreaseCapacity(Length + numberOfPos);
+
+            for (int i = Length - 1; i >= position; i--)
+            {
+                _array[i + numberOfPos] = _array[i];
+            }
+            Length += numberOfPos;
+
+            // write the default values ​​in the vacated cells
+            for (int i = position; i < position + numberOfPos; i++)
+            {
+                _array[i] = default(T);
+            }
+        }
+
+        /// <summary>
+        /// The method shifts the elements of the array(right of the specified position),
+        /// starting from the specified position, to the left by the specified number of positions.
+        /// (Elements shifted to the left of the array are removed)
+        /// </summary>
+        private void LeftShift(int position, int numberOfPos)
+        {
+            for (int i = position; i < Length; i++)
+            {
+                if (i - numberOfPos >= 0)
+                {
+                    _array[i - numberOfPos] = _array[i];
+                }
+            }
+
+            Length -= numberOfPos;
+        }
+
+        /// <summary>
+        /// Возможность ручного изменения значения Capacity с сохранением уцелевших данных
+        /// (данные за пределами новой Capacity сохранять не нужно).
+        /// </summary>
+        public void SetCapacity(int newCapacity) // 2*
+        {
+            if (newCapacity < 0)
+                throw new ArgumentException("Argument must be positive");
```

Comment:
Not just "argument", exception message shouldn't be too generic. Here you can call it what it is - "capacity must be greater than 0"